Speed up deserialization

bal-e opened the pull request:
It looks like the `kmip-ttlv` deserialization code is a serious bottleneck during signing in Cascade.  This PR attempts to resolve this by adding a manual deserializer implementation.

ximon18 committed:
WIP

bal-e committed:
Add module 'ttlv'

This commit imports some draft optimization work intended for
'kmip-ttlv', with the goal of deprecating that crate entirely.  The fast
scanner and formatter defined here should provide good performance and
have a simple API, that 'src/de/mod.rs' needs to be rebased onto.

bal-e committed:
[types] Derive 'Ordinalize' for all enums

bal-e committed:
[types/common] Impl manual serde

The only breaking change is to 'CryptographicUsageMask', defining it via
'bitflags' instead of 'enum-flags'.  The latter hasn't been updated in a
while and 'bitflags' is the preferred crate for this purpose.

bal-e committed:
[types/request] Impl manual serde

Also expand 'impl_ttlv_serde' with struct support.

bal-e committed:
[types/response] Impl manual serde

ximon18 commented on src/de/mod.rs:
```diff
@@ -0,0 +1,549 @@
+use std::{
```
This whole file can be deleted, right?

ximon18 replied:
I removed it.

ximon18 committed:
Update request serialization tests to use the new Formatter.

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,394 @@
+//! Formatting TTLV elements really fast.
```
```suggestion
//! Serializing TTLV elements really fast.
```

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+//! Scanning TTLV elements really fast.
```
```suggestion
//! Deserializing TTLV elements really fast.
```

ximon18 replied:
Also, if this only supports deserializing of requests at the moment we should make that clear.

(ignore me, the "only request" aspect is to do with which types have had fast scan support implemented for them or not, not an issue with the deserializer itself)

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+/// A fast TTLV scanner.
+///
+/// In most cases, errors are highly unlikely.  This scanner optimizes for
+/// this by eliding error handling and pinpointing facilities -- it detects
```
```suggestion
/// this by omiting error handling and pinpointing facilities -- it detects
```

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+
+/// A fast TTLV scanner.
+///
+/// In most cases, errors are highly unlikely.  This scanner optimizes for
```
Note: I have often encountered parsing issues, not at the TTLV level, but when expecting parsing to result in a specific structure and the input not matching that. I suggest that this statement be removed and instead talk about lack of error handling and error location determination as a means to increase performance, not because it isn't likely to be needed.

bal-e replied:
From our perspectives as developers, that is true.  But the code is under development and we are actively looking for problems.  Once this code works and we've moved on, errors really will be rare.

ximon18 replied:
I thought so too but 5 years on parsing new structures or using ones less well tested in the past I still ran into new problems, or even hit issues when trying a new HSM for the first time. Anyway, leave it if you wish, I think it's optimistic.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+/// this by eliding error handling and pinpointing facilities -- it detects
+/// errors without tracking any details of them.  In some cases, it will
+/// delay errors, returning garbage data instead of [`Err`].  If scanning
+/// with [`FastScanner`] fails, a different scanner needs to be used to
```
```suggestion
/// with [`FastScanner`] fails, use the slower `kmip_ttlv::de::from_slice()` to 
```

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+/// with [`FastScanner`] fails, a different scanner needs to be used to
+/// ascertain the position and cause of the error.
+#[derive(Clone)]
+#[must_use = "Call 'FastScanner::finish()' to finish scanning"]
```
How does this `must_use` work? If I do `let a = FastScanner::new(buf);` then I have "used" the struct so the lint will not be violated, correct? And it thus has nothing to do with `finish()` which the `must_use` text refers to?

bal-e replied:
See <https://doc.rust-lang.org/stable/reference/attributes/diagnostics.html#the-must_use-attribute>.  The same mechanism is used to lint `Result` and `Option`.  Yes, in your example the lint would no longer be violated, but then you'd get the unused-variable lint.

You're right that the message isn't quite right.  For `Formatter`, I have a better way to enforce completion, using the `FormatDone` value; but that isn't fool-proof either.  I'll update the message, but AFAICT there isn't an ideal solution for enforcing `finish()`.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+impl<'a> FastScanner<'a> {
+    /// Construct a new [`FastScanner`].
+    pub fn new(input: &'a [u8]) -> Result<Self, FastScanError> {
+        let num_blocks = input.len() / 8;
```
Add a comment here explaining that section 9.1.1.3 of the KMIP 1.0 specification declares that a valid TTLV byte sequence is always a multiple of 8 bytes.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+        let num_blocks = input.len() / 8;
+        if !input.len().is_multiple_of(8) {
+            // The input contains a partial block.
+            return Err(FastScanError::assert());
```
I find `assert()` a strange name for this functionality. Maybe instead `::with_backtrace()` ?

bal-e replied:
A backtrace is only included in debug builds; it's a no-op for release builds.  I named it `assert` as the caller is asserting that something has gone wrong.  This pattern can be found in other codebases, but there doesn't seem to be a single conventional name for the function.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+    /// Construct a new [`FastScanner`].
+    pub fn new(input: &'a [u8]) -> Result<Self, FastScanError> {
+        let num_blocks = input.len() / 8;
+        if !input.len().is_multiple_of(8) {
```
As this check does not use `num_blocks` it could be done one line higher.

bal-e replied:
Sure, assuming you're noting this for style rather than performance.

ximon18 replied:
Yes, it's like a wonky painting, once I saw it I can't unsee it now :-)

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+
+impl<'a> FastScanner<'a> {
+    /// Construct a new [`FastScanner`].
+    pub fn new(input: &'a [u8]) -> Result<Self, FastScanError> {
```
Would it make sense to reject zero length slices?

bal-e replied:
That's a good question.  There isn't a strict need to, and you can sometimes end up with a `FastScanner` with empty data (e.g. when parsing a struct with zero fields).  Disallowing it would only create new edge cases for callers.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+        Ok(Self::from_blocks(input))
+    }
+
+    /// Construct a new [`FastScanner`] from a block slice.
```
Is "block slice" a known / defined term? It's not familiar to me and I didn't find it in a Google search.

bal-e replied:
Nope, feel free to suggest a better name.  Maybe just "slice of 8-byte blocks" is good.

ximon18 replied:
Maybe this shouldn't be here at all as a named function but instead as a From impl? Then we don't have to worry about finding a good name for it ;-)

bal-e replied:
I'd rather it be an explicit method because it's important to read the documentation and understand how to use it; while you can add documentation on a `From` impl, it's less likely to be read.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+#[must_use = "Call 'FastScanner::finish()' to finish scanning"]
+pub struct FastScanner<'a> {
+    /// The currently remaining input.
+    remaining: &'a [[u8; 8]],
```
This is a clever trick, I hadn't realized that ALL KMIP TTLV sequences are ALWAYS composed of 8 byte chunks.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+    }
+
+    /// Whether the scanner is finished.
+    pub const fn is_empty(&self) -> bool {
```
If the scanner were given an empty slice this would return true but can one say that it is "finished" if it had no work to do? If `new()` rejected empty slices I would then be inclined to call this `is_finished()` and the ambiguity would be gone.

bal-e replied:
Let's continue that discussion on your comment on `new()`.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+    /// This should only be checked if the consistency of the previously
+    /// scanned data is important, e.g. if acting on it.  It can be ignored
+    /// while parsing.
+    pub const fn delayed_error(&self) -> bool {
```
Since you use `is_` above should this be `has_delayed_error()`? When would "consistency of the previously scanned data" not be important? If it can be ignored while parsing why be able to check it while parsing? Doesn't the user learn this anyway by calling `finish()` and getting an error back? I'm unclear what the purpose of this function is. Maybe this is for internal use and should be private?

bal-e replied:
Yes, for consistency this should be `has_delayed_error()`.  Will fix.

`FastScanner::scan_int()` doesn't return `Err` for every invalid input.  Sometimes, it will just return a garbage `i32` and record a delayed error.  This is useful for eliminating branch conditions (replacing them with some cheap branchless arithmetic), but it means that the caller cannot assume that a particular `i32` value they read (e.g. a batch count) was what the source of the TTLV data actually intended.  In cases where they want that guarantee, e.g. to allocate space upfront (although it is a bad idea), they can check this first.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+    /// The length of the next element will be checked; if it is entirely loaded
+    /// in the buffer, `true` is returned.  Otherwise, or if there is no next
+    /// element, `false` is returned.
+    pub fn have_next(&self) -> bool {
```
Should this be private?

bal-e replied:
This was deliberately made public, and is not used privately.  It was a late addition, and allows callers to distinguish a "not enough data in your buffer" error from a "your data is invalid" error.  If a caller is reading TTLV data from an I/O stream, they can call this function to determine whether they have buffered enough data to parse it all, or whether they need to fetch more.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+        self.failure_word != 0
+    }
+
+    /// Whether a whole element is available.
```
What is an "element"? You don't introduce that term above.

The KMIP 1.0 specification says mixes up the terms "data object", "field" and "item" so I admit it is not a good source of naming inspiration. Your "element" seems to be 8 bytes long so, assuming we're not in the V part of a single TTLV "item", those 8 bytes represent the TTL part (a 3 byte tag, a 1 byte type, and a 4 byte length) which one could refer to as the "header" (and indeed in `scan_struct()` you refer to `expected_header`.

So maybe instead "Check if the next TTLV sequence is entirely present in the buffer by parsing the TTL header and checking that the buffer has at least L bytes (of V) remaining"  (also avoiding the term "loaded" which could be misinterpreted to imply that "loading" is done somewhere by the scanner)?

Also, this function doesn't appear to be used anywhere and I'm wondering what use it has? `is_empty()` will tell you if you haven't finished yet, and calling `fast_scan()` on a type that was augmented using the macro will consume everything that can be consumed according to the data structure being deserialized into.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+    /// in the buffer, `true` is returned.  Otherwise, or if there is no next
+    /// element, `false` is returned.
+    pub fn have_next(&self) -> bool {
+        let [head, rest @ ..] = self.remaining else {
```
Maybe use `header` instead of `head` like you do in `scan_struct()` below?

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+        };
+
+        let length = u32::from_be_bytes(head[4..8].try_into().unwrap());
+        let available = rest.as_flattened().len();
```
Is it cheaper to multiply or shift than call flattened?

bal-e replied:
Doesn't matter; LLVM is great.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+        let length = u32::from_be_bytes(head[4..8].try_into().unwrap());
+        let available = rest.as_flattened().len();
+
+        if usize::MAX as u32 == u32::MAX {
```
Can't you compare `usize::BITS` and `u32::BITS` here instead as both are u32?

bal-e replied:
Oooh, that's a much better way of going about this!

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+    /// If there is remaining input, or an error was delayed, [`Err`] is
+    /// returned.  If additional remaining input is allowed, use
+    /// [`Self::finish_non_exhaustive()`] instead.
+    pub fn finish(self) -> Result<(), FastScanError> {
```
The caller cannot tell if the call failed because more input is available or because there was an error.

Perhaps if they knew there was more input available they would try and scan more, e.g. another message? Or should they call `have_next()` if they want that ability?

Also, you say in the `must_use()` at the top that `finish()` has to be called, but removing the call to `finish()` doesn't break anything that I can see.

And calling `finish()` isn't even needed if the buffer contains a complete KMIP message and one uses the macro defined `fast_scan()` function to deserialize an entire `RequestMessage`.

I find this API a bit confusing.

bal-e replied:
1. This is deliberate.  The caller can check for more input using `is_empty()`, or ignore the extra input using `finish_non_exhaustive()`.  `have_next()` is not the right tool for that, it is unnecessarily expensive.

2. `finish()` _is_ important to call.  Without it, delayed errors get ignored.

3. The macro calls `finish()` for you as it's easy to forget and is boilerplate.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+    /// `struct`.
+    #[inline]
+    pub fn scan_struct(&mut self, tag: Tag) -> Result<Self, FastScanError> {
+        let expected_header = (tag.value() as u64) << 40 | 0x01_00000000;
```
What is this doing? This line even fails to compile in Rust Playground, but that turns out to be because the type is inferred by Rust to be u64 from later lines of code, but on its own that inference is not available.

A Tag contains a u32 value and contains a 24 bit, 3 byte KMIP TTLV "tag" value. You treat it as a u64, i.e. 8 bytes, so presumably that results in the left most bytes being all zero, then you shift it left by 40 bits, i.e. by 5 bytes so that would move the tag value bits to the left most end of the 64-bit value. You then OR it with what presumably is a 5-byte 64-bit value (but it's not annotated as u64) setting a 1 at the right of the 4th byte.

You then XOR this with the actual header received, then AND it with a mask that will preserve the 4 left bytes and the rightmost 3 bits. If the expected and actual header had the same tag and type (the top most 4 bytes) those would cancel out and all be zero. The lower 4 bytes (the actual received L value) would remain unmodified. The AND operation will clear everything in the lower 4 bytes, i.e. most of the received L value, except the lower 3 bits...? Why.

And `failure_word`: I assume this is "word" in the machine architecture sense, but then what size "word" are we referring to here? You will set the `failure_word` if the expected tag or type don't match the received tag or type, or if the lower 3 bits of the received L are ARE set you consider that a "failure". Ah, you are checking here if the received L is not a multiple of 8 then it will have one or more of the lower 3 bits set, which is invalid because all lengths must be a multiple of 8.

Isn't this wrong given this part of the KMIP 1.0 spec:

> If the Item Type is Integer, Enumeration, Text String, Byte String, or Interval, then the Item Length is the number of bytes excluding the padding bytes

The received L value is thus allowed to be less than 8 for some types, right?

bal-e replied:
Yes, this is concise code doing a lot of work.  Regarding the last part of your comment: this is `scan_struct()`, other functions use different length checks.  The length field of a struct always includes padding, and so is always a multiple of 8.

ximon18 replied:
This concise code needs documenting then. I had indeed forgotten that I was looking at struct specific parsing!

ximon18 reviewed (commented):
Review still in progress, but I have to stop for now.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+            return Err(FastScanError::assert());
+        }
+        let (header, rest) = self.remaining.split_first().unwrap();
+        let header = u64::from_ne_bytes(*header);
```
I see `from_be_bytes()` above but `from_ne_bytes()` below. Should they all be the same?

bal-e replied:
No, they were chosen quite carefully.  Some of the `u64`s around here are represented in big-endian, and others are native-endian.  I tried to use native-endian where possible.  I should add comments pointing these out, they can be quite subtle.

ximon18 reviewed (commented):
A few replies and additional comments.

ximon18 committed:
FIX: The Cryptographic Parameters attribute value deserializes to a rich type, not a string.

ximon18 committed:
FIX: Missing space in Object Type attribute name.

ximon18 committed:
FIX: Contact Information should be a Text String, not its own type.

ximon18 committed:
Extract common test code into a test util mod, and assert that constructed wire requests can be deserialized correctly.

ximon18 committed:
Remove the half-finished early manual deserializer work.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+    /// The currently remaining input.
+    remaining: &'a [[u8; 8]],
+
+    /// A failure word.
```
I assume this is "word" in the machine architecture sense, but then what size "word" are we referring to here? I think it would be good to either rename this to "error_occurred" or "has_error" or something, or to explain why it is called "word".

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+            return Err(FastScanError::assert());
+        }
+
+        let ptr = input.as_ptr().cast::<[u8; 8]>();
```
```suggestion
        // SAFETY: We just checked that the input is a multiple of 8 bytes.
        let ptr = input.as_ptr().cast::<[u8; 8]>();
```

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+        }
+    }
+
+    /// The remaining input to the scanner.
```
Is it worth documenting here that because some errors don't cause scanning to stop immediately that you can't assume that `remaining()` points to the input that could not be deserialized?

bal-e replied:
Yes, I'll make a note about that.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+    ///
+    /// This should only be checked if the consistency of the previously
+    /// scanned data is important, e.g. if acting on it.  It can be ignored
+    /// while parsing.
```
```suggestion
    /// Call this function if you want to use scanned data prior to calling
    /// [`finish()`].
    ///
    /// To optimize the scanner, some kinds of error will not cause the
    /// scanner to fail until [`finish()`] is called, and so until that point it
    /// is possible that the data returned by prior scanning may be
    /// incorrect.
    ///
    /// This function checks for an error now, rather than waiting until
    /// [`finish()`] is invoked.
```

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+        self.failure_word |= (expected_header ^ header) & 0xFFFFFFFF_00000007;
+        let length = header as u32 as usize / 8;
+        let (body, rest) = rest.split_at_checked(length).ok_or(FastScanError::assert())?;
+        self.remaining = rest;
```
If `self.failure_word` is not checked here in order to avoid branching, how is it okay to use the `?` operator, won't that also cause branching?

bal-e replied:
The goal is to minimize branching.  Unfortunately, length-related checks can't be avoided.  I have been thinking about potential improvements, but I think they'd be out of scope.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+    ///
+    /// The returned sub-scanner can be used to scan the fields of the
+    /// `struct`; `self` will independently continue from the end of the
+    /// `struct`.
```
As all `scan` functions return a sub-scanner and leave `self` at the next position in the input after the encountered TTLV item, maybe say it once at the top somewhere instead of repeating it for each function?

bal-e replied:
It is only used for the `scan_struct` functions; I prefer making it easier to find for users, even at the cost of some boilerplate.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+///   error, but may be localized to the particular kind of input message.
+///
+/// The caller can distinguish incompatibilities from the lower-level errors
+/// by re-parsing the input using [`SlowScanner`].  In order to distinguish
```
Note that the SlowScanner doesn't exist yet?

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1009 @@
+/// - Crash.  This is ideal if the error prevents the caller from fulfilling
+///   critical functionality.
+///
+/// The caller should re-parse the input using [`SlowScanner`]; if the error
```
Note that the SlowScanner doesn't exist yet?

bal-e committed:
Handle review feedback

- Document delayed errors and streaming operation in FastScanner.
- Factor out the FastScanner's bit-arithmetic into helper functions.
- Use the try operator in roundtrip-ttlv tests.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1166 @@
+    /// Construct a new [`FastScanner`] from non-exhaustive input.
+    ///
+    /// This should be preferred over [`Self::new()`] if more input may be
+    /// appended in the future (e.g. during streaming operation).  It will
```
How would one append such extra input?

bal-e replied:
e.g. using a `Vec` as input, appending to it, and creating a new scanner.  I can note that down.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1166 @@
+
+        // Load the header without changing its endianness, so that it and
+        // 'expected' have matching byte layouts.
+        let actual = u64::from_ne_bytes(*header);
```
I'm afraid I'm missing something here. How is it correct to XOR a `to_be()` (big endian) value with a `from_ne_bytes()` (native endian) value?

bal-e replied:
TTLV uses big-endian.  So `*header` looks like `42 00 00 01 00 00 00 08` (tag, type, then length). `u64::from_be_bytes(*header)` would yield `0x4200000100000008u64`.  `u64::from_be_bytes(*header)` is equivalent to `u64::from_be(u64::from_ne_bytes(*header))`, i.e. `u64::from_be_bytes(*header).to_be()` is equivalent to `u64::from_ne_bytes(*header)`.

bal-e replied:
Consider `from_ne_bytes()` as loading an integer _without changing its endianness_ instead of as loading an integer _in native endianness_.

ximon18 replied:
If one looks at the [documentation](https://doc.rust-lang.org/std/primitive.usize.html#method.from_ne_bytes) for `from_ne_bytes()` it states the the resulting value will be in native endian format, i.e. it could be little endian or it could be big endian, whatever the native endiannness used by the machine is. It even warns for portability to NOT use native endianness but instead to explicitly use big endian or little endian. So I'm still confused about the mixing of both `ne` and `be` together.

bal-e replied:
`from_ne_bytes()` will load data in native endianness _if the source is in native endianness_, which it assumes.  In our case, the source we're loading from is in big endian; since `from_ne_bytes()` doesn't change the endianness of the data it loads, it's ending up loading a big endian value.

ximon18 replied:
That's a terribly named and documented function, IMO. It basically does NOTHING, except tell the type checker to close its eyes. So, if I understand correctly, the endianness here is whatever endianness the source data that was put in memory at that location had, it has nothing to do with the native platform architecture. I'd be inclined to write my own function that does the unsafe call to transmute and give it a better name, like `as_raw_bytes()` or something.

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,398 @@
+    /// `struct`; once it is finished, `self` can continue.
+    #[inline]
+    pub fn format_struct(&mut self, tag: Tag) -> Result<FieldFormatter<'_>, TruncationError> {
+        let buffer = &mut self.buffer[self.position..];
```
If `buffer` here has the type `&mut [MaybeUninit<[u8; 8]>` that is the same as the type returned by `fn remaining()` so why does one use `get_unchecked_mut()` and the other use slice syntax?

bal-e replied:
Oh, I thought that was a mistake on my end, but it's for good reason!  `self.remaining()` would borrow `&mut self`, which would conflict with the later borrow of `&mut self.position`.  By accessing `.buffer` directly, the compiler understands that we only wanted this field, and that the two mutable borrows are independent.

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,398 @@
+    #[inline]
+    pub fn format_big_int(&mut self, tag: Tag, value: &[u8]) -> FormatResult {
+        let header = tag.with_type(Type::BigInteger).value() as u64;
+        let length = u32::try_from(value.len()).unwrap() as u64;
```
So if the caller provides a slice more than 4 GiB long this will panic? I suppose that's a pretty unlikely case but as you are happy to return a `TruncationError` below it seems strange not to return an error here - or not to document the panic scenario?

bal-e replied:
I'll note it in the documentation.

bal-e replied:
I've taken a slightly neater approach: `Formatter::new()` will automatically truncate the input to 4GiB.  That way, this length conversion will never fail.

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,398 @@
+    pub fn format_big_int(&mut self, tag: Tag, value: &[u8]) -> FormatResult {
+        let header = tag.with_type(Type::BigInteger).value() as u64;
+        let length = u32::try_from(value.len()).unwrap() as u64;
+        let length = (length + 7) & !7; // round up to 8 bytes
```
Neat trick.

bal-e replied:
Oh... I should actually replace that with `.next_multiple_of()`.  Clippy should have warned about that too, I know they have a lint for it.

bal-e committed:
Handle review feedback, round 2

bal-e committed:
[fast_scan] Use 'Type'

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1172 @@
+    ///
+    /// Returns `true` if it does.
+    #[inline(always)]
+    fn test_ttl(&self, header: &[u8; 8], tag: Tag, r#type: Type, length: u32) -> Option<()> {
```
Why does this return an `Option<()>` instead of `bool`?

bal-e replied:
Oh, it seems I changed things midway.  I'll have a look.

ximon18 replied:
I think this has happened because elsewhere you have a `test_tt_8l()` that returns `Option<usize>`, so for consistency you have `test_ttl()` that also returns `Option`.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1172 @@
+    }
+
+    #[inline(always)]
+    fn calc_tt_8l(header: &[u8; 8], tag: Tag, r#type: Type) -> (u64, usize) {
```
A function level RustDoc comment here would be good, denoting what the two return values represent.

bal-e replied:
I've documented the conventions on the `impl` block.

ximon18 replied:
Hmm, I meant say something about what the u64 is or means, not that it returns a u64 as the user can see that :-) Anyway, it'll be fine, it can be worked out by seeing how it is used by the callers.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1172 @@
+        // we'll include the non-'usize' bits in the mask, and so ensure that
+        // they are zero in the actual header.
+        //
+        // This is 0xFFFF_0007 on 16-bit, else 0x0000_0007.
```
I'm afraid I don't understand why this distinction is being made for 16-bit platforms.

bal-e replied:
It explains what the constant will evaluate to.  The reason for `0xFFFF_0007` is that we don't want to support lengths greater than 65536 on 16-bit platforms; `FastScanner` will always parse the whole element in one go (from contiguous memory) and such a large element would simply never fit.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1172 @@
+    }
+
+    #[inline(always)]
+    fn calc_tt(header: &[u8; 8], tag: Tag, r#type: Type) -> (u64, usize) {
```
A function level RustDoc comment here would be good, denoting what the two return values represent.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1166 @@
+        // We don't need to support lengths that can't fit in 'usize'; such
+        // objects wouldn't fit in memory.  If 'usize' is smaller than 'u32',
+        // we'll include the non-'usize' bits in the mask, and so ensure that
+        // they are zero in the actual header.
```
For the 16-bit case you say you "ensure that they are are zero in the actual header" but this  code doesn't affect the "actual header" at all.

Perhaps what you mean is that you ensure that this function won't return a length larger than the system can support and so any caller of this function that uses the returned length to constrain how many bytes of data it consumes will limit itself to the safe reduced number of bytes.

However if you are limiting to the maximum that will fit in memory it still won't fit as memory will be used by other things, not just this data. Further, won't this break parsing because instead of the caller skipping to the actual end of the TTLV it will stop short of the end of the TTLV and attempt to read another one at the too early position in the stream?

So I'm still confused :-)

bal-e replied:
If the length exceeds the limit specified here, it sets a delayed error.  Yes, you could miss it if you're using the `_opt()` family of functions and then call `finish_non_exhaustive()`.  But I have chosen not to seriously consider that issue.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1166 @@
+            // We know 'u32' fits in a 'usize', so we can cast directly.
+            (mismatch, length as usize)
+        } else {
+            // Generate a mask for the length.
```
I'm inclined to think we should not support systems where `usize` is smaller than 32 bits.

bal-e replied:
I don't see the point.  It only adds a few more lines of code, and they don't affect 32-bit and 64-bit systems.

ximon18 replied:
I guess I just don't like this code and would rather not have it especially since the systems we are targeting will be 64-bit systems.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1172 @@
+    #[inline]
+    pub fn scan_fixed_struct(&mut self, tag: Tag, length: u32) -> Result<Self, FastScanError> {
+        debug_assert!(length.is_multiple_of(8));
+        if self.remaining.len() < 1 + length as usize / 8 {
```
Would it be beneficial to store the result of the `length as usize / 8` calculation as it is done again below?

bal-e replied:
With regards to performance: nope, it's trivial.  LLVM knows they end up as the same thing.

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1172 @@
+        self.check_ttl(header, tag, Type::Boolean, 8);
+        // Load the value without changing its endianness.
+        let value = u64::from_ne_bytes(*value);
+        // Check that it all bits except the lowest are zero.
```
```suggestion
        // Check that all bits except the lowest are zero.
```

ximon18 commented on src/ttlv/fast_scan.rs:
```diff
@@ -0,0 +1,1172 @@
+        };
+        // Load the value without changing its endianness.
+        let value = u64::from_ne_bytes(*value);
+        // Check that it all bits except the lowest are zero.
```
```suggestion
        // Check that all bits except the lowest are zero.
```

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+    /// Construct a new [`Formatter`].
+    ///
+    /// The input will be truncated to the first 4GiB, and to a multiple of 8
+    /// bytes, as per limitations of the TTLV format.  It should ideally be
```
Why should it ideally be aligned? What happens if it is not?

bal-e replied:
Better performance.  Will note.

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+    /// `Formatter::from_raw_parts(buffer, position)` is sound if and only if:
+    /// - `position <= buffer.len()`.
+    /// - `buffer[..position]` is initialized.
+    /// - `buffer.len() <= 2^32`.
```
Here you say "<= 2^32" but in the code it says "<= 1<<32" ?

bal-e replied:
... yes, the two are equivalent.

ximon18 replied:
Ah, sorry, misread that as both using ^ or both using <<, my bad.

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+        }
+
+        let header = tag.with_type(Type::BigInteger).value();
+        // NOTE: 'buffer.len() * 8 <= 2^32' so 'num_blocks * 8 < 2^32'.
```
Should this say "buffer.len() <= 2^32" ?

bal-e replied:
Nope, the other one was wrong.  Fixed.

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+                    .copy_from_nonoverlapping(value.as_ptr(), value.len())
+            };
+        } else {
+            // The value is empty; we don't write anything.
```
Does this leave the bytes uninitialized?

bal-e replied:
No; since `value` was empty, zero blocks will be written out for it.

ximon18 replied:
And if no bytes are written to the MaybeUninit that is okay? (I'm not familiar with MaybeUninit)

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+        }
+
+        let header = tag.with_type(Type::TextString).value();
+        // NOTE: 'buffer.len() * 8 <= 2^32' so 'num_blocks * 8 < 2^32'.
```
Should this say "buffer.len() <= 2^32" ?

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+            unsafe {
+                ptr.sub(value.len())
+                    .copy_from_nonoverlapping(value.as_ptr(), value.len())
+            };
```
If (and that's a big IFF) I understand this correctly:
- `as_mut_ptr_range().end()` returns a pointer to the byte AFTER the last byte of the given range.
- The given range is `buffer[1..]` which ends at the end of `buffer`.
- As `buffer` is set above to `self.buffer[self.position..]` the end of `buffer` is the end +1 of the entire buffer given to the formatter.
- `ptr.sub(value.len())` will move the pointer to the end of the formatter buffer + 1 back by `value.len()` bytes then `copy_from_nonoverlapping()` will copy bytes into the formatter buffer at that point.

Why does this appear to be writing from the end of the _*entire*_ formatter buffer minus the value length, instead of from the current write position in the formatter buffer onwards?

What did I miss?

bal-e replied:
... you're right, I messed up.  And this wasn't caught by `roundtrip_ttlv` because I forgot to make it generate big integers to test :facepalm:

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+        }
+
+        let header = tag.with_type(Type::ByteString).value();
+        // NOTE: 'buffer.len() * 8 <= 2^32' so 'num_blocks * 8 < 2^32'.
```
Should this say "buffer.len() <= 2^32" ?

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+
+//----------- FieldFormatter -------------------------------------------------
+
+/// A [`Formatter`] for filling a particular TTLV field.
```
It's a bit strange to link to `Formatter` saying this is one when it is not, because actually it is a `FieldFormatter`?

ximon18 replied:
Ah, I see, it is a thin wrapper around a `Formatter`, so in that sense it _is_, effectively and by deref, a `Formatter`.

bal-e replied:
Yep!

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+/// A [`Formatter`] can delegate to a [`FieldFormatter`] to format a TTLV
+/// field such as a structure.  This provides an incremental API to fill the
+/// field.
+#[must_use = "Complete the formatter using `.finish()`"]
```
Didn't we say for the scanner that a `must_use` on the struct wasn't the right way to use `must_use`?

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+
+    /// The header slot.
+    ///
+    /// The length of the created field will be written here.
```
The tag and type are assumed to be already written here, right?

bal-e replied:
Yes, will note.

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+
+impl FieldFormatter<'_> {
+    /// Successfully finish formatting.
+    pub fn finish(self) -> FormatDone {
```
Should/could this be a Drop impl instead?

bal-e replied:
Unfortunately not.  Consider a function using `FieldFormatter`; if it encounters an error, it would never get around to calling `.finish()`, but it would still call `.drop()`.  `.finish()` is deliberately designed to be called only if the formatter finishes successfully.  This ensures, should an error occur at any point during formatting, all the intermediate written bytes are truncated.

ximon18 commented on src/ttlv/format.rs:
```diff
@@ -0,0 +1,434 @@
+//----------- FormatResult ---------------------------------------------------
+
+/// The result of formatting a field.
+pub type FormatResult = Result<FormatDone, TruncationError>;
```
I'm curious, why `FormatDone` and not `()` ?

bal-e replied:
I've added documentation on `FormatDone` to explain.

ximon18 commented on src/ttlv/types.rs:
```diff
@@ -0,0 +1,240 @@
+
+    /// A TTLV date-time.
+    ///
+    /// This is a 64-bit signed UNIX time value, in units of seconds.
```
```suggestion
    /// This is a 64-bit signed POSIX time value, in units of seconds.
```

bal-e replied:
The convention is "Unix", e.g. https://en.wikipedia.org/wiki/Unix_time.

ximon18 replied:
Okay, but https://docs.oasis-open.org/kmip/spec/v1.2/kmip-spec-v1.2.html#_toc8707 says:

> Date-Time values are POSIX Time values encoded as Long Integers.

ximon18 commented on src/ttlv/types.rs:
```diff
@@ -0,0 +1,240 @@
+    /// This corresponds to [`Duration`], but has a limited range and a
+    /// granularity of seconds.
+    ///
+    /// [`Duration`]: std::time::Duration
```
I can't find anywhere else in the code that uses `Duration`...?

bal-e replied:
Yes, I was just noting the conceptual similarity between the two.

ximon18 replied:
Oh, I thought you meant that the user would be able to get a `Duration` out of your code.

ximon18 commented on src/ttlv/types.rs:
```diff
@@ -0,0 +1,240 @@
+
+impl fmt::Display for TagType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{} {}", self.r#type(), self.tag())
```
Why write type then tag, when this is a TagType (tag before type) and also when in TTLV the first two T's are Tag and Type, not type and tag? And maybe this would be better written as "tag (type)" to make it clearer that the type belongs to the tag?

bal-e replied:
Because "Struct 0x420078" seemed more readable than "0x420078 (Struct)".  It is only called a `TagType` because that matches their order in memory.  I'd prefer to leave this alone for now and return once a slow-scanner with tag mapping is implemented here; until then, nobody is really examining the output.

ximon18 commented on src/ttlv/types.rs:
```diff
@@ -0,0 +1,240 @@
+        type Strategy = prop::strategy::Map<Range<u32>, fn(u32) -> Tag>;
+
+        fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
+            (0..1 << 24).prop_map(Tag::new)
```
What is this doing?

bal-e replied:
Generating a random integer between 0 and 2^24 and converting it into a tag.  Which would be the expected behavior for randomly generating tags.

ximon18 commented on src/ttlv/types.rs:
```diff
@@ -0,0 +1,240 @@
+        type Strategy = prop::strategy::Map<Range<u8>, fn(u8) -> Type>;
+
+        fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
+            (1..11).prop_map(|x| unsafe { core::mem::transmute(x) })
```
Is this maybe better written as `(0x01..0x0A)` to match the enum above?

ximon18 commented on src/types/common.rs:
```diff
@@ -258,6 +409,60 @@ impl Display for KeyMaterial {
+            | KeyFormatType::Opaque
+            | KeyFormatType::PKCS1
+            | KeyFormatType::PKCS8
+            | KeyFormatType::X509 => scanner.scan_bytes(Self::TAG).map(|s| Self::Bytes(s.into())),
```
Should X509 here be ECPrivateKey?

ximon18 replied:
Also, this is for parsing tag 0x420042 I think, not 0x420043.

ximon18 replied:
Also should this call `finish()`?

bal-e replied:
Good catch.  Yes, 0x420042 is `KeyFormatType` which is provided by the caller.  `finish()` only needs to be called on struct scanners created by the current function, not on the scanner provided as a parameter.  I would have documented that convention, but I don't want to put too much effort into it because I'd rather rewrite it later.

ximon18 commented on src/types/common.rs:
```diff
@@ -258,6 +409,60 @@ impl Display for KeyMaterial {
+
+    pub fn format(&self, formatter: &mut Formatter<'_>) -> FormatResult {
+        match self {
+            KeyMaterial::Bytes(v) => formatter.format_bytes(Self::TAG, v),
```
I think this is wrong as it will write tag 0x420043 when it should write 0x420042.

ximon18 replied:
Also, should it call `finish()`?

bal-e replied:
My understanding is that 0x420042 is `KeyFormatType`, not `KeyMaterial`.  That was my reading of the serde attributes too.

ximon18 replied:
Right, sorry got confused by the code comment above which I suspect mistakenly references 43 instead of 42.

ximon18 commented on src/types/common.rs:
```diff
@@ -258,6 +409,60 @@ impl Display for KeyMaterial {
+                TransparentDHPublicKey::fast_scan(scanner).map(Self::TransparentDHPublicKey)
+            }
+
+            _ => todo!(),
```
If we get a response containing legal but unsupported key material I don't think we want to panic. Instead I think we should skip the structure and either return variant Structure with empty bytes or do as the original code did which was to return variant Structure with the V bytes of the structure in the vec.

ximon18 commented on src/types/common.rs:
```diff
@@ -258,6 +409,60 @@ impl Display for KeyMaterial {
+            KeyMaterial::TransparentDHPrivateKey(this) => this.format(formatter),
+            KeyMaterial::TransparentDHPublicKey(this) => this.format(formatter),
+
+            KeyMaterial::Structure(_) => todo!(),
```
I think we can live with this for now, but should it be `unimplemented!()` instead?

bal-e replied:
Sure, will fix.

ximon18 commented on src/types/common.rs:
```diff
@@ -266,6 +471,24 @@ pub struct TransparentSymmetricKey {
 
+impl TransparentSymmetricKey {
+    pub const TAG: Tag = Tag::new(0x420043);
+    pub const KEY_TAG: Tag = Tag::new(0x42003F);
```
Why does this refer to 0x42003F here?

bal-e replied:
As per the spec, `TransparentSymmetricKey` is defined as a struct containing a `Key` field.  `Key` uses tag 0x42003F, I just double-checked this in the KMIP 1.0 spec.

ximon18 replied:
Ah the bug is in the original code, a missing serde rename attribute on the key field.

ximon18 commented on src/types/common.rs:
```diff
@@ -280,6 +503,33 @@ pub struct TransparentDSAPrivateKey {
 
+impl TransparentDSAPrivateKey {
+    pub const TAG: Tag = Tag::new(0x420043);
+    pub const P_TAG: Tag = Tag::new(0x42005E);
```
Interesting. I think this is correct and the original code looks wrong as it lacks the tag numbers for the inner byte fields.

bal-e replied:
Yeah, I have no clue how that worked.

ximon18 replied:
It didn't work, it has just never been used.

ximon18 commented on src/types/common.rs:
```diff
@@ -751,66 +1228,139 @@ impl CryptographicDomainParameters {
     }
 }
 
+impl CryptographicDomainParameters {
```
Why is this not defined like the impl for `CryptographicParameters`?

bal-e replied:
Because the `q_length` field is defined as an `i32`, and does not provide `fast_scan()` etc. (after all, it doesn't know what tag to scan with).  If it were changed to a custom `QLength` type, it could indeed use the `impl_ttlv_serde` macro, but I was striving to avoid API changes.

ximon18 commented on src/types/common.rs:
```diff
@@ -751,66 +1228,139 @@ impl CryptographicDomainParameters {
+bitflags::bitflags! {
+    /// See KMIP 1.0 section 3.14 [Cryptographic Usage Mask](https://docs.oasis-open.org/kmip/spec/v1.0/os/kmip-spec-1.0-os.html#_Toc262581188).
+    #[derive(Clone, Copy, PartialEq, Eq)]
+    pub struct CryptographicUsageMask: i32 {
```
Why is this changed from an enum to a struct?

bal-e replied:
An `enum` was never the right representation for it, because it is a bitmask and the previous "enum variants" could be combined together.  `bitflags` is the de facto crate for representing bitfields in Rust and it (IMO quite correctly) represents them as structs.

ximon18 replied:
I thought the whole point of bit flags is to combine them together, and in the case of a usage mask it can have multiple bits set at once. Anyway, if the reason is because of the way `bitflags` works then let's leave it like this.

ximon18 commented on src/types/mod.rs:
```diff
@@ -7,7 +7,321 @@
 //!
 //! The attributes on the Rust types are used by the `kmip-ttlv` crate to guide the (de)serialization correctly to/from
 //! the KMIP binary TTLV format.
```
```suggestion
//! Rust types defined by this module are specially augmented to provide (de)serialization support.
//!   - Serde crate attributes provide support for using the "old" Serde based (de)serializer which is powered by the `kmip-ttlv` crate. This support can be accessed via the `from_slice()` and `to_vec()` functions.
//! 
//!   - `fast_scan()`/`format()` functions, mostly generated using the `impl_ttlv_serde!` macro defined below, offer a new faster (de)serialization implementation.
//! 
//! Note: Support for the new fast (de)serializer is currently limited to request related KMIP types, (de)serializing response types must still be done using the "old" Serde based (de)serializer. Also not yet implemented is support for a new fallback "slow" (de)serializer capable of reporting useful information about (de)serialization errors. The "old" (de)serializer can be used as a fallback but may not encounter the same (de)serialization failures as the "new" fast (de)serializer.
```

ximon18 commented on src/types/common.rs:
```diff
@@ -266,6 +471,24 @@ pub struct TransparentSymmetricKey {
+    pub fn fast_scan(scanner: &mut FastScanner<'_>) -> Result<Self, FastScanError> {
+        let mut scanner = scanner.scan_struct(Self::TAG)?;
+        let key = scanner.scan_bytes(Self::KEY_TAG)?.into();
+        scanner.finish()?;
```
Why is it correct to call `.finish()` here? If this struct was being deserialized from a larger TTLV sequence the scanner will not be empty and so calling `finish()` will cause an error. What am I missing?

bal-e replied:
Note that we called `.finish()` on the output of `.scan_struct()`, which is limited in scope to `TransparentSymmetricKey`.

ximon18 commented on src/types/mod.rs:
```diff
@@ -7,7 +7,321 @@
+        impl $type {
+            pub const TAG: Tag = Tag::new($tag);
+
+            pub fn fast_scan(scanner: &mut FastScanner<'_>) -> Result<Self, FastScanError> {
```
The naming is a bit inconsistent here. Why fast scan but not fast format? Why say fast at all?

bal-e replied:
Because I intended there to be a `SlowScanner` which uses pretty much the same infrastructure.  Formatting doesn't need a slow/fast distinction because it has a simple failure case that doesn't need elaboration.

ximon18 commented on src/types/mod.rs:
```diff
@@ -7,7 +7,321 @@
+
+            fn fast_scan_inner(mut $scanner: FastScanner<'_>) -> Result<Self, FastScanError> {
+                let this = $fast_scan;
+                $scanner.finish()?;
```
Same as I reported in `common.rs`: Why is it correct to call .finish() here? If this struct was being deserialized from a larger TTLV sequence the scanner will not be empty and so calling finish() will cause an error. What am I missing?

bal-e replied:
We are calling `.finish()` here on a **sub-**scanner, which is scoped to a particular struct (whatever this is being implemented on).

ximon18 replied:
How do you know that you are not receiving the top level scanner that is processing a `RequestMessage` and is intended to be re-used for the next request message to be received as well? Wouldn't this break?

bal-e replied:
Yes, the top-level scanner needs to be handled a little differently; but that can be implemented at the top-level, using the normal `fast_scan()` functions, then calling `.finish()` manually.  We don't have any implementations of that top-level scanning here, so you don't see it.

ximon18 commented:
@bal-e: Would it make sense to provide Scan/Format traits that `impl_ttlv_serde!()` and manual implementations impl? Currently the consistency of the various implementations is not guaranteed as they don't have to conform to a specified interface.

ximon18 commented on src/types/mod.rs:
```diff
@@ -7,7 +7,321 @@
+    };
+
+    // Impl TTLV (de)serialization for a boolean wrapper.
+    (bool $type:ident as $tag:literal) => {
+        impl $type {
+            pub const TAG: Tag = Tag::new($tag);
+
+            pub fn fast_scan(scanner: &mut FastScanner<'_>) -> Result<Self, FastScanError> {
+                Self::fast_scan_with(scanner, Self::TAG)
+            }
+
+            pub fn fast_scan_with(scanner: &mut FastScanner<'_>, tag: Tag) -> Result<Self, FastScanError> {
+                scanner.scan_bool(tag).map(Self)
+            }
+
+            pub fn fast_scan_opt(scanner: &mut FastScanner<'_>) -> Result<Option<Self>, FastScanError> {
+                Self::fast_scan_opt_with(scanner, Self::TAG)
+            }
+
+            pub fn fast_scan_opt_with(scanner: &mut FastScanner<'_>, tag: Tag) -> Result<Option<Self>, FastScanError> {
+                scanner.scan_opt_bool(tag).map(|s| s.map(Self))
+            }
+
+            pub fn format(&self, formatter: &mut Formatter<'_>) -> FormatResult {
+                self.format_with(formatter, Self::TAG)
+            }
+
+            pub fn format_with(&self, formatter: &mut Formatter<'_>, tag: Tag) -> FormatResult {
+                formatter.format_bool(tag, self.0)
+            }
+        }
+    };
```
Could this impl be simplified by invoking `impl_serde_ttlv!` again, something like:

```suggestion
    (bool $type:ident as $tag:literal) => {
        impl_ttlv_serde!(struct $type as $tag {
          fast_scan = |scanner| scanner.scan_bool($tag).map(Self),
          format = |&self, formatter| formatter.format_bool($tag, self.0)
        }
    };
```

bal-e replied:
No, because the `struct` impl calls `.scan_struct()` and `.format_struct()` for you.

ximon18 commented on src/types/request.rs:
```diff
@@ -180,11 +187,28 @@ impl Attribute {
+        let name = AttributeName::fast_scan(&mut scanner)?;
+        let index = AttributeIndex::fast_scan_opt(&mut scanner)?;
+        let value = AttributeValue::fast_scan(&mut scanner, &name)?;
+        Self(name, index, value)
```
Why is it correct to call `finish()` [here](https://github.com/NLnetLabs/kmip-protocol/blob/df775f509b8ded3b5e5002291cc60e1608b4ef27/src/types/common.rs#L481) but not in this code?

bal-e replied:
Because this one is using the macro which inserts `.finish()` for you.

ximon18 commented on src/types/request.rs:
```diff
@@ -456,6 +641,31 @@ pub struct KeyValue(
     #[serde(skip_serializing_if = "Option::is_none")] pub Option<Vec<Attribute>>,
 );
 
+impl KeyValue {
```
Why isn't this done with the complex struct form of the `impl_ttlv_serde!()` macro with the `#[option+vec]` support?

bal-e replied:
Because it takes an additional `KeyFormat` parameter.

ximon18 commented on src/types/response.rs:
```diff
@@ -48,18 +78,64 @@ pub struct KeyValue {
+        let Some(mut scanner) = scanner.scan_opt_struct(Self::TAG)? else {
+            return Ok(None);
+        };
+        let key_material = KeyMaterial::fast_scan(&mut scanner, format)?;
```
Couldn't you just call `Self::fast_scan()` here?

bal-e replied:
No, because `fast_scan` calls `scan_struct()`.

ximon18 commented on src/types/response.rs:
```diff
@@ -48,18 +78,64 @@ pub struct KeyValue {
 pub struct TemplateAttribute {
     #[serde(rename = "0x420043")]
-    pub name: KeyMaterial,
+    pub names: Option<Vec<Name>>,
```
👀

ximon18 commented on tests/roundtrip_ttlv.rs:
```diff
@@ -0,0 +1,431 @@
+        any::<u32>().prop_map(FieldData::Interval),
+    ];
+
+    field(leaf_data.prop_recursive(3, 27, 3, |inner| {
```
A comment about why the numbers 3, 27 and 3 were chosen for depth, desired size and expected branch size would be useful here.

ximon18 commented on tests/roundtrip_ttlv.rs:
```diff
@@ -0,0 +1,431 @@
+    ];
+
+    field(leaf_data.prop_recursive(3, 27, 3, |inner| {
+        prop::collection::vec(field(inner), 0..10).prop_map(|fields| FieldData::Structure(fields.into_boxed_slice()))
```
Likewise for the range 0..10 here.

ximon18 commented:
I've now looked at all of the changes, and added comments/questions. Regarding `tests/roundtrip_ttlv.rs` I'm not familiar with proptest so I haven't really reviewed it, but I get the idea and it seems okay, and as it found an issue when you mistakenly used the wrong KMIP type in one function it must be doing something useful! :-)

I would approve this, the only concern I have is about the, to me, seemingly inconsistent invocation of `.finish()` which I referred to in a few comments.

I also haven't tried actually using this library from client code, though I believe you have tested it with kmip2pkcs11. Is that correct? However, if we find issues when integrating I don't expect them to be due to the architecture of what you have built, but in the detail of the scan/format impls for a particular type, and we can just fix those as we find them, if any.

I also didn't comment on the fact that the docs for the crate haven't been updated, but as this merges into `next` we can deal with that when we eventually feel ready to make a new release of this crate.

bal-e committed:
Handle review feedback, round 3

ximon18 reviewed (approved):
There are still some open comments but none of them are worth blocking this PR on, approving!
